Recovery behaviors

orduno opened the pull request:
## Basic Info
| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | #49 |
| Primary OS tested on | Ubuntu 16.04 |
| Primary platform tested on | Gazebo simulation of turtlebot |
--- 
## Description of contribution in a few bullet points
Developed an example of how to implement recovery behaviors on the behavior tree based navigator (`bt_navigator`).
* Extended the navigation behavior tree to detect if the robot is stuck and take corrrective actions before pursuing further navigation. 
* Implemented a simple algorithm for detecting if the robot is stuck. Created the corresponding `is_stuck` condition behavior tree node.
* Created `stop`, `back_up` and `spin` action nodes and defined the corresponding `task_client`s and `task_server`s.
* Defined a `nav2_behaviors` node to host the `task_server`s for the multiple behaviors. Implemented robot behaviors for spinning and backing up.
* Implemented dummy planner and controller for testing.
--- 
## Future work that may be required in bullet point

* We might want to have the behavoirs `task_server` not as a separate package but part of the robot.
* Multiple improvements are suggested (see `TODO`s) for the algorithms used for detecting if the robot is stuck, for spinning and backing up.

mjeronimo commented on nav2_behaviors/include/nav2_behaviors/back_up.hpp:
```diff
@@ -0,0 +1,53 @@
+  nav2_tasks::TaskStatus backUp(const nav2_tasks::BackUpCommand::SharedPtr command);
+
+protected:
+  double min_linear_vel_, max_linear_vel_, linear_acc_lim_, goal_tolerance_distance_;
```
Anticipating doxygen, I'd put these on separate lines.

mjeronimo commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+
+  robot_ = std::make_unique<nav2_robot::Robot>(temp_node);
+
+  vel_pub_ =
```
Should use the robot class.

orduno replied:
Changed.

mjeronimo commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // TODO(orduno): For now, backing up is time based
+    auto duration = 3s;
```
Could be a parameter contained in the BackUpCommand.

mhpanah replied:
I assume there is a task to check for collision before backing up.   If we don't have that task we should not initiate this task.

mjeronimo commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,195 @@
+
+  robot_ = std::make_unique<nav2_robot::Robot>(temp_node);
+
+  vel_pub_ =
```
Robot class.

orduno replied:
Fixed

mjeronimo commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,195 @@
+    }
+
+    // TODO(orduno): For now, spinning is time based
+    auto duration = 4s;
```
Could be a parameter in the command message.

orduno replied:
See previous comment in `back_up.cpp`

mjeronimo commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,195 @@
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = vel;
+
+    vel_pub_->publish(cmd_vel);
```
robot_->sendVelocity(cmd_vel)

orduno replied:
Fixed

mjeronimo commented on nav2_bt_navigator/CMakeLists.txt:
```diff
@@ -17,6 +17,8 @@ find_package(nav2_tasks REQUIRED)
 find_package(nav2_msgs REQUIRED)
 find_package(nav2_robot REQUIRED)
 find_package(behavior_tree_core REQUIRED)
+find_package(tf2 REQUIRED)
```
Why does the BT Navigator required additional packages? Should just be updating the XML and registering new types.

orduno replied:
It should, the tf2 is used by `nav2_tasks::spin_action` I tried adding it to the CMake of `nav2_tasks` however I get some linking errors. Can you help me take a look?

mhpanah commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+    }
+
+    // Output control command
+    geometry_msgs::msg::Twist cmd_vel;
```
We should move the backup velocity commends to Robot class.

orduno replied:
Agree, fixed.

mhpanah commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,195 @@
+    cmd_vel.linear.x = 0.0;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = 0.5;
+    vel_pub_->publish(cmd_vel);
```
Let's use  Robot's class  `sendVelocity` method.

orduno replied:
@mhpanah fixed
@SteveMacenski see previous comment

mjeronimo commented on nav2_dummy_controller/src/dummy_controller.cpp:
```diff
@@ -0,0 +1,104 @@
+
+  auto temp_node = std::shared_ptr<rclcpp::Node>(this, [](auto) {});
+
+  vel_pub_ =
```
Use a robot object.

orduno replied:
yes, we need to standarize the output of the controllers, dwb is not using the robot. I've opened an issue: #381

mjeronimo commented on nav2_dummy_controller/src/dummy_controller.cpp:
```diff
@@ -0,0 +1,104 @@
+  return TaskStatus::SUCCEEDED;
+}
+
+void DummyController::setZeroVelocity()
```
Should be a method of the robot class. Otherwise, we get this functions scattered around the code. We have a similar function, for example, in the DWB controller. Should all be factored out into one place (the robot class).

orduno replied:
Agree, filed #381 to follow-up with both controllers.

mjeronimo commented on nav2_dummy_planner/src/dummy_planner.cpp:
```diff
@@ -0,0 +1,71 @@
+
+  RCLCPP_INFO(get_logger(), "Found a dummy path");
+
+  nav2_tasks::ComputePathToPoseResult result;
```
It would be helpful to actually return some path data, even in the dummy planner case. This at least helps us to check that the path is being communicated properly from the planner to the controller via the BT classes.

orduno replied:
Yes, filed an issue #382

mjeronimo commented on nav2_tasks/include/nav2_tasks/is_stuck_condition.hpp:
```diff
@@ -0,0 +1,149 @@
+  {
+    RCLCPP_INFO(get_logger(), "IsStuckCondition::constructor");
+
+    vel_cmd_sub_ = this->create_subscription<geometry_msgs::msg::Twist>(
```
These are both available via the robot class.

orduno replied:
The dwb controller is currently direclty publishing the velocity and need to change that so this condition node works correctly. Opened an issue #383.

mhpanah commented on nav2_dummy_controller/src/dummy_controller.cpp:
```diff
@@ -0,0 +1,104 @@
+// Copyright (c) 2018 Intel Corporation
+//
```
What is the point of dymmy controller?  Do we need it in master?

SteveMacenski replied:
I second this

orduno replied:
@mhpanah @SteveMacenski I've found it really useful having dummy modules for development (FakeLocalization, Dummy controller & planner). What problem do you see with having these?

SteveMacenski replied:
It might be useful, but its not a production tool. If you wanted to include a dummy controller for CI or development, then it should go into the system tests package and not outside as an existing package. Fake localization is a needed tool to do full simulations in gazebo of an actual robotics system, where this is a placeholder to make testing and evaluation easier but not a production-level test or simulation.

mhpanah replied:
I’m not convinced that putting this dummy controller and planner on the master adds any value.   If this code segment is meant to be used as an example, then it could perhaps be explained in readme files or if it is for development then as @SteveMacenski said it could either go into the system tests package or on a different branch

orduno replied:
Agree on moving to System Tests.

mjeronimo commented on nav2_tasks/include/nav2_tasks/is_stuck_condition.hpp:
```diff
@@ -0,0 +1,149 @@
+
+    using namespace std::chrono_literals;
+
+    if (current_velocity_ == nullptr) {
```
Another benefit of using the robot class is that this kind of code isn't replicated in several places; there is one place that manages the pubs/subs.

orduno replied:
Agree, see comment above.

mhpanah commented on nav2_dummy_planner/include/nav2_dummy_planner/dummy_planner.hpp:
```diff
@@ -0,0 +1,40 @@
+// Copyright (c) 2018 Intel Corporation
```
Do we need this in master?

orduno replied:
See previous comment.

crdelsey commented:
1. This change set is very big. Can we break it into smaller chunks or get a design doc to go with it?
2. How does this compare to the ROS 1 behavior?
3. Since this is all new code, it should have some unit tests to go with it.

SteveMacenski commented on nav2_behaviors/CMakeLists.txt:
```diff
@@ -0,0 +1,80 @@
+cmake_minimum_required(VERSION 3.5)
+project(nav2_behaviors)
```
I would rename this to `nav2_recovery_behaviors` unless you're intending to support other non-recovery behaviors here

orduno replied:
This will host all sorts of behaviors, used for recovery or some other non-recovery task, i.e. backing up could be used as a recovery action or for docking the robot.

SteveMacenski replied:
The ROS-y way of doing this isnt to make 1 large package but many smaller packages. I would encourage us to make a distinction between recovery and social/other behaviors that are event, opportunistic based and not as a result of planning/control failure. 

With that said, I generally eye larger packages with many themed executables, but the point stands that I think a distinction between recoveries and social behaviors is necessary as the way they're used and how they're to be triggered will differ dramatically.

SteveMacenski commented on nav2_behaviors/include/nav2_behaviors/back_up.hpp:
```diff
@@ -0,0 +1,53 @@
+namespace nav2_behaviors
+{
+
+class BackUp : public rclcpp::Node
```
Shouldn't there be an implementation class for all the recoveries to follow? Allows some structure or common code amongst recoveries as well as making meta-recovery nodes that have a configurable pointer vector of recoveries.

orduno replied:
Agree, I'll push a commit shortly that has a Behavior base clase. BackUp and Spin derive from it. I like the idea of meta-recoveries, and that's the idea behind identifying the minimum set of primitives from which we can compose more complex behaviors.

SteveMacenski commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,136 @@
+  auto start_time = std::chrono::system_clock::now();
+  auto time_since_msg = std::chrono::system_clock::now();
+
+  while (true) {
```
while not-control-Ced (ok())? while true would not accept signals

SteveMacenski commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,136 @@
+    cmd_vel.linear.x = -0.05;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = 0.0;
+    vel_pub_->publish(cmd_vel);
```
There's no call to check if the path is clear in the costmap, this would be dangerous

orduno replied:
Agree. For now we're backing up a minimum distance < .5 robot diameter, but still, we should check. Created #379 to follow-up.

SteveMacenski commented on nav2_behaviors/src/main.cpp:
```diff
@@ -0,0 +1,40 @@
+
+  rclcpp::executors::SingleThreadedExecutor exec;
+
+  auto spin = std::make_shared<nav2_behaviors::Spin>();
```
Making all behaviors derive from a shared class would allow someone to spin over a pointer vector as another way of architecting this. This also falls in line with the ROS1 implementation of nav_core recoveries.

orduno replied:
Agree, will push a change shortly.

SteveMacenski commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,198 @@
+  auto start_time = std::chrono::system_clock::now();
+  auto time_since_msg = std::chrono::system_clock::now();
+
+  while (true) {
```
while rclcpp::ok()

orduno replied:
Fixed

SteveMacenski commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,198 @@
+    cmd_vel.linear.x = 0.0;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = 0.5;
+    vel_pub_->publish(cmd_vel);
```
Checking costmaps?

orduno replied:
Created an issue #379 to follow-up with a separate PR.

SteveMacenski commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,198 @@
+    geometry_msgs::msg::Twist cmd_vel;
+    cmd_vel.linear.x = 0.0;
+    cmd_vel.linear.y = 0.0;
+    cmd_vel.angular.z = vel;
```
This rotational velocity should probably scale with angular distance so that it doesnt abruptly start and stop or cause massive wheel slippage when accelerating

orduno replied:
Some scaling is done in 153 considering the acceleration limit and distance left.

orduno commented on nav2_behaviors/include/nav2_behaviors/spin.hpp:
```diff
@@ -0,0 +1,60 @@
+
+  std::unique_ptr<nav2_tasks::SpinTaskServer> task_server_;
+
+  std::shared_ptr<rclcpp::Publisher<geometry_msgs::msg::Twist>> vel_pub_;
```
Thanks, I've made the change.

orduno commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+  RCLCPP_INFO(get_logger(), "Initializing the BackUp behavior");
+
+  // TODO(orduno) Pull values from param server or robot
+  max_linear_vel_ = 0.0;
```
Totally agree, we need to do some more work on the Robot class. I've filed two issues (#337 and #338) to follow-up later.

orduno commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+  // Start listening for incoming BackUp task requests
+  task_server_->startWorkerThread();
+
+  RCLCPP_INFO(get_logger(), "Initialized the BackUp  behavior");
```
good catch!

orduno commented on nav2_behaviors/src/back_up.cpp:
```diff
@@ -0,0 +1,134 @@
+  task_server_->setExecuteCallback(std::bind(&BackUp::backUp, this, std::placeholders::_1));
+
+  // Start listening for incoming BackUp task requests
+  task_server_->startWorkerThread();
```
Yeah, we'll be able to do some pretty cool stuff once we move to managed nodes :)

orduno commented on nav2_tasks/include/nav2_tasks/stop_action.hpp:
```diff
@@ -0,0 +1,137 @@
+namespace nav2_tasks
+{
+
+class StopAction : public BT::ActionNode, public rclcpp::Node
```
OK, i'll take a look.

orduno commented:
> Spin and backup classes are using a lot of similar codes. 

@mhpanah Agree both classes use smiliar code, will push a change that defines a Behavior base class that solves this. About the second question:

> These two classes can be combined into one. We can have a Recovery class that contains spin and backup method.

Having a single class (Recovery) containing both methods would require us to edit and recompile the class itself. Trying to follow the Open-Closed principle.

SteveMacenski commented:
@orduno I'd recommend naming that base class with the word `recovery` somewhere in it. Through out this PR we've been interchanging `recovery` and `behavior` which I wouldnt want to be considered interchangeable. The HRI folks would not consider these to be behaviors but recovery actions. I think calling them recovery behaviors is fair but leaving the generalized `behavior` name open for future growth of the project to contain actual HRI-type behaviors

orduno commented:
Guys, I've addressed most of your feedback. I've opened issues for the tasks that I think are not critical and could involve sufficient code changes to merit separate PRs.

mhpanah commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,174 @@
+      quaternion.x, quaternion.y, quaternion.z, quaternion.w)).getEulerYPR(yaw, pitch, roll);
+}
+
+bool Spin::getRobotYaw(double & yaw)
```
I don't think this method is necessary. tf2::getYaw does this operation.
`yaw = tf2::getYaw(current_pose.pose.orientation)`

orduno replied:
Thanks, changed

mhpanah commented on nav2_behaviors/src/spin.cpp:
```diff
@@ -0,0 +1,174 @@
+  return TaskStatus::RUNNING;
+}
+
+void Spin::getAnglesFromQuaternion(
```
Same with below.  tf2::getEulerYPR does this operation.

orduno replied:
Changed.

mhpanah commented on nav2_behaviors/include/nav2_behaviors/spin.hpp:
```diff
@@ -0,0 +1,62 @@
+    const geometry_msgs::msg::Quaternion & quaternion,
+    double & yaw, double & pitch, double & roll);
+
+  bool getRobotYaw(double & yaw);
```
Please see my comment below.

orduno replied:
changed

mhpanah commented on nav2_behaviors/include/nav2_behaviors/spin.hpp:
```diff
@@ -0,0 +1,62 @@
+
+  nav2_tasks::TaskStatus controlledSpin();
+
+  void getAnglesFromQuaternion(
```
Please see my comment below.

orduno replied:
changed

orduno commented:
> @orduno I'd recommend naming that base class with the word `recovery` somewhere in it. Through out this PR we've been interchanging `recovery` and `behavior` which I wouldnt want to be considered interchangeable. The HRI folks would not consider these to be behaviors but recovery actions. I think calling them recovery behaviors is fair but leaving the generalized `behavior` name open for future growth of the project to contain actual HRI-type behaviors

@SteveMacenski Good point, I wasn't thinking out of the navigation realm when naming it `behavior`. I was thinking more like `navigation_behavior` but even then this class is not meant to implement robot conduct.

The intent of this class is to represent basic motions or primitives that any type of movement task, recovery or not, can use. For that reason I'd like to ommit using `recovery`.

How about a name along the lines of `basic_motions` or `motion_primitives`?

mhpanah commented on nav2_tasks/include/nav2_tasks/stop_action.hpp:
```diff
@@ -0,0 +1,142 @@
+
+    // TODO(orduno) What should this node return for a long running action
+
+    // Give some time to the robot to get stable
```
Please remove unused code.

orduno replied:
Thanks for pointing this out. I needed to discuss with @mjeronimo what a long running action should return when calling `tick()` on a derived class from `BT::ActionNode`. I was previously returning `RUNNING` however that sometimes failed. Mike clarifed that the overridden `tick()` is really what the worker thread is doing on each cycle and not equivalent to the tick concept of a BT node. This should only return success or failed once it's finished doing it's work.

mhpanah commented on nav2_tasks/include/nav2_tasks/spin_action.hpp:
```diff
@@ -0,0 +1,59 @@
+    // Rotate 90deg CCW
+    quaternion.setRPY(0, 0, M_PI / 2);  // yaw, pitch and roll are rotation in z, y, x respectively
+    // quaternion.normalize();
+
```
How about `command_->quaternion = tf2::toMsg(quaternion);`?  Then delete lines 47-53

orduno replied:
You're right.

mjeronimo commented on nav2_tasks/include/nav2_tasks/stop_action.hpp:
```diff
@@ -0,0 +1,142 @@
+    // RCLCPP_INFO(get_logger(), "tick:: completed");
+    // new_call = true;
+
+    auto sleep_time = 5s;
```
Sleeping in a tick function will pause the whole tree. Instead of sleeping you can monitor the elapsed time.

orduno replied:
Changed with the new refactored code.

orduno commented on nav2_motion_primitives/include/nav2_motion_primitives/motion_primitive.hpp:
```diff
@@ -0,0 +1,156 @@
+
+    RCLCPP_INFO(node_->get_logger(), "MotionPrimitive ran for %.2f seconds", elapsed_seconds.count());
+
+    return status;
```
That's something we need to discuss further #411, for now I've pushed a change to the `motion_primitive` base class to send a stop command to the robot after executing the primitive.

orduno commented on nav2_motion_primitives/src/back_up.cpp:
```diff
@@ -0,0 +1,112 @@
+BackUp::BackUp(rclcpp::Node::SharedPtr & node)
+: MotionPrimitive<nav2_tasks::BackUpCommand, nav2_tasks::BackUpResult>(node)
+{
+  // TODO(orduno) Pull values from param server or robot
```
Removed the tolerance, in this case we don't need it, it's just a small motion.
Yes, the rest of the attributes should be pulled from the robot. Once the robot class is ready we'll have quite some refactoring to do #378.

orduno commented on nav2_tasks/include/nav2_tasks/stop_action.hpp:
```diff
@@ -0,0 +1,142 @@
+  {
+    // A BT node can't get values from the blackboard in the constructor since
+    // the BT library doesn't set the blackboard until after the tree if build.
+    if (!initialized_) {
```
As mentioned, I'm directly deriving from `BT:ActionNode` and not the `BtActionNode` class. However after our discussion of how stop works, I think it makes sense to create a server that will stop the robot and take actions to stabilize it's motion. Just pushed some changes for that.

mjeronimo commented on nav2_bt_navigator/behavior_trees/parallel_w_recovery.xml:
```diff
@@ -0,0 +1,35 @@
+      </FallbackStar>
+      <SequenceStar name="navigate">
+        <ComputePathToPose endpoints="${endpoints}" path="${path}"/>
+        <Parallel threshold_M="1">
```
With the new upstream BT code, "Parallel" is now "ParallelNode" and "threshold_M" is now just "threshold"

orduno replied:
Fixed

mjeronimo commented on nav2_bt_navigator/src/navigate_to_pose_behavior_tree.cpp:
```diff
@@ -30,6 +35,12 @@ NavigateToPoseBehaviorTree::NavigateToPoseBehaviorTree(rclcpp::Node::SharedPtr n
+  factory_.registerNodeType<nav2_tasks::StopAction>("Stop");
+  factory_.registerNodeType<nav2_tasks::BackUpAction>("BackUp");
+  factory_.registerNodeType<nav2_tasks::SpinAction>("Spin");
+
```
When you rebase, make sure that we're not registering the Parallel node. This is now done by the BT library (he's fixed this issue).

orduno replied:
OK
